### ISSUE TITLE
Add automated SQL generation accuracy evaluation harness

### ISSUE THREAD

**Issue by Krooted** [Krooted/thrive-ui#57]
## Summary

Build an automated evaluation script that measures text-to-SQL generation accuracy against a ground-truth question/SQL dataset, enabling before/after comparison when RAG training changes are made.

---

## Motivation / Problem Statement

We've invested heavily in RAG enrichment — auto-enhance schema (column stats, type-mismatch detection, implicit relationships), multi-hop JOIN path discovery, view definition extraction, and LLM-generated table documentation. However, we have **no systematic way to measure whether these changes actually improve SQL generation accuracy**. 

Current test coverage measures RAG *retrieval* quality (top-1 accuracy, MRR in `test_milvus_vs_chroma.py`) but nothing measures whether the final *generated SQL* is correct. Without an eval harness, we're flying blind on the impact of training pipeline changes.

---

## Proposed Solution / High-level Design

- **Core idea** – A CLI-invokable eval script that reads a ground-truth CSV, runs each question through `generate_sql()`, executes both generated and expected SQL against the database, compares result sets, and outputs a scorecard.

- **Key components**:
  1. **Ground-truth CSV** (`eval/ground_truth.csv`) — curated question/SQL pairs with category tags
  2. **Eval runner** (`eval/run_eval.py`) — orchestrates the evaluation loop
  3. **Result comparator** — compares generated vs expected SQL output (execution accuracy)
  4. **Scorecard reporter** — outputs per-category and aggregate metrics as a markdown table or JSON

- **Data flow**:
  ```
  ground_truth.csv → eval runner → generate_sql() per question
       ↓                                    ↓
  execute expected SQL              execute generated SQL
       ↓                                    ↓
  compare result sets ← ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
       ↓
  scorecard (markdown / JSON)
  ```

---

## User Stories / Acceptance Criteria

| User Story | Acceptance Criteria |
|------------|---------------------|
| As a developer, I want to run an eval suite before and after RAG training changes so that I can quantify accuracy impact | 1. Script reads a ground-truth CSV and produces a scorecard <br> 2. Scorecard includes per-category breakdown (joins, aggregations, filters, etc.) <br> 3. Results are saved to a timestamped output file for comparison |
| As a developer, I want execution-based accuracy (not just string matching) so that semantically equivalent SQL is counted as correct | 1. Both generated and expected SQL are executed against the database <br> 2. Result sets are compared (row-count match + value match) <br> 3. Partial credit is tracked (correct columns, wrong filter, etc.) |
| As a developer, I want to easily expand the ground-truth set so that eval coverage grows over time | 1. CSV format is simple: `question`, `expected_sql`, `category`, `difficulty` <br> 2. Categories include: `simple_lookup`, `aggregation`, `join`, `multi_join`, `date_filter`, `type_sensitive`, `view_query` <br> 3. At least 20-30 seed questions covering each category |

---

## Non-Functional Requirements

- **Performance** – Full eval suite (30 questions) should complete in under 5 minutes
- **Security** – Eval script must respect the existing `forbidden_references.json` security validator; ground-truth SQL must not reference forbidden tables
- **Reproducibility** – Results must be deterministic given the same training data and ground-truth set (set LLM temperature to 0 for eval runs)
- **Isolation** – Eval runs should be read-only against the database (SELECT only in ground-truth SQL)

---

## Dependencies / Related Work

- [x] `auto_enhance_schema()` — column stats, type mismatches, implicit relationships (committed)
- [x] `_find_join_paths()` — multi-hop JOIN path discovery (committed)
- [x] `train_ai_documentation()` — LLM-generated table docs (committed)
- [x] `train_all()` 4-step pipeline (in progress, uncommitted)
- [ ] Ground-truth question set must be created per target database — cannot be generic

---

## Test Plan / QA Checklist

- **Unit tests** – Test result-set comparator handles: identical results, column reordering, row reordering, empty results, NULL handling
- **Integration tests** – Run eval against a known database with 5 questions where expected results are pre-verified
- **Manual testing** – Verify scorecard output is readable and categories are correctly assigned

---

## Potential Risks / Trade-offs

| Risk | Impact | Mitigation |
|------|--------|------------|
| Ground-truth SQL is database-specific | Eval set not portable across deployments | Document that each deployment needs its own ground-truth CSV |
| LLM non-determinism | Scores vary between runs even with same training | Set temperature=0 for eval; run 3x and average if needed |
| Execution accuracy misses semantic errors | SQL returns right row count but wrong data | Include value-level comparison, not just row counts |
| Large eval sets slow down iteration | Developers skip running evals | Keep seed set small (20-30); allow category filtering via CLI flag |

---

## Metrics to Track

| Metric | Description |
|--------|-------------|
| **Execution accuracy** | % of questions where generated SQL returns the same result set as expected |
| **Valid SQL rate** | % of questions where generated SQL is syntactically valid and executes without error |
| **Category accuracy** | Execution accuracy broken down by question category |
| **Error analysis** | For failures: classify as wrong table, wrong join, wrong filter, wrong aggregation, syntax error |

---

## Release Notes / Changelog (draft)

- Added automated SQL generation evaluation harness (`eval/run_eval.py`)
- Added ground-truth CSV template with 20+ seed questions across 7 categories
- Scorecard outputs per-category accuracy, valid SQL rate, and error classification

**Comment by univibe25** [comment]
Closing as implemented. The repository now has the automated agent evaluation harness under `evals/`, `scripts/agent_eval_harness.py`, report generation, representative question matrices, optional judging, and an authenticated durable evaluation workspace with review history.
